File: frontend/src/api/notationVariete.ts
import client from './client'

// ── Types ─────────────────────────────────────────────────────────────────────

export interface NotationRead {
  id_notation: number
  nom_variete: string
  breeder?: string | null
  date_notation?: string | null

  // Partie A — Culture /30
  vigueur_sante?: number | null
  productivite_structure?: number | null
  soif?: number | null

  // Partie B — Consommation /70
  apparence_structure?: number | null
  profil_aromatique?: number | null
  saveur_qualite?: number | null
  effet_puissance?: number | null

  // Labo
  taux_thc?: number | null
  taux_cbd?: number | null
  terpene_dominant?: string | null
  commentaire_labo?: string | null
  notes_generales?: string | null

  // Calculés
  total_culture?: number | null
  total_consommation?: number | null
  note_finale?: number | null

  created_at?: string | null
  updated_at?: string | null
}

export type NotationCreate = Omit<NotationRead, 'id_notation' | 'total_culture' | 'total_consommation' | 'note_finale' | 'created_at' | 'updated_at'>

export type NotationUpdate = Partial<NotationCreate>

// ── Types stats extraction ────────────────────────────────────────────────────

export interface ExtractionStat {
  avg_rosin_pct: number | null
  nb_rosin: number
  avg_hash_pct: number | null
  nb_hash: number
}

/** Clé = nom_variete */
export type ExtractionStatsMap = Record<string, ExtractionStat>

// ── API ────────────────────────────────────────────────────────────────────────

export const notationVarieteAPI = {
  list: () =>
    client.get<NotationRead[]>('/notations/').then(r => r.data),

  get: (id: number) =>
    client.get<NotationRead>(`/notations/${id}`).then(r => r.data),

  create: (payload: NotationCreate) =>
    client.post<NotationRead>('/notations/', payload).then(r => r.data),

  update: (id: number, payload: NotationUpdate) =>
    client.put<NotationRead>(`/notations/${id}`, payload).then(r => r.data),

  delete: (id: number) =>
    client.delete(`/notations/${id}`),

  getExtractionStats: () =>
    client.get<ExtractionStatsMap>('/notations/utils/extraction-stats').then(r => r.data),

  exportCsv: () => {
    window.open(`${client.defaults.baseURL ?? ''}/notations/export/csv`, '_blank')
  },
}

File: backend/app/schemas/notation_variete.py
"""Schémas Pydantic pour NotationVariete (Classement des variétés)."""
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, Field


# ── Champs communs ─────────────────────────────────────────────────────────────

class NotationBase(BaseModel):
    nom_variete:            str
    breeder:                Optional[str]   = None
    date_notation:          Optional[date]  = None

    # Partie A — Culture (/30)
    vigueur_sante:          Optional[float] = Field(None, ge=0, le=10)
    productivite_structure: Optional[float] = Field(None, ge=0, le=10)
    soif:                   Optional[float] = Field(None, ge=0, le=10)

    # Partie B — Consommation (/70)
    apparence_structure:    Optional[float] = Field(None, ge=0, le=15)
    profil_aromatique:      Optional[float] = Field(None, ge=0, le=15)
    saveur_qualite:         Optional[float] = Field(None, ge=0, le=20)
    effet_puissance:        Optional[float] = Field(None, ge=0, le=20)

    # Labo (informatif)
    taux_thc:               Optional[float] = None
    taux_cbd:               Optional[float] = None
    terpene_dominant:       Optional[str]   = None
    commentaire_labo:       Optional[str]   = None
    notes_generales:        Optional[str]   = None


class NotationCreate(NotationBase):
    pass


class NotationUpdate(BaseModel):
    """Tous les champs optionnels pour la mise à jour partielle."""
    nom_variete:            Optional[str]   = None
    breeder:                Optional[str]   = None
    date_notation:          Optional[date]  = None

    vigueur_sante:          Optional[float] = Field(None, ge=0, le=10)
    productivite_structure: Optional[float] = Field(None, ge=0, le=10)
    soif:                   Optional[float] = Field(None, ge=0, le=10)

    apparence_structure:    Optional[float] = Field(None, ge=0, le=15)
    profil_aromatique:      Optional[float] = Field(None, ge=0, le=15)
    saveur_qualite:         Optional[float] = Field(None, ge=0, le=20)
    effet_puissance:        Optional[float] = Field(None, ge=0, le=20)

    taux_thc:               Optional[float] = None
    taux_cbd:               Optional[float] = None
    terpene_dominant:       Optional[str]   = None
    commentaire_labo:       Optional[str]   = None
    notes_generales:        Optional[str]   = None


class NotationRead(NotationBase):
    id_notation:   int
    created_at:    Optional[datetime] = None
    updated_at:    Optional[datetime] = None

    # Scores calculés (retournés par l'API)
    total_culture:      Optional[float] = None   # /30
    total_consommation: Optional[float] = None   # /70
    note_finale:        Optional[float] = None   # /100

    model_config = {"from_attributes": True}
